src/calliope/preprocess/model_definition.py

```diff
@@ -153,3 +183,80 @@ def _log_overrides(init_model_def: AttrDict, overriden_model_def: AttrDict) -> N
         else:
             continue
         LOGGER.debug(message)
+
+
+class TemplateSolver:
+    """Resolves templates before they reach Calliope models."""
+
+    TEMPLATES_SECTION: str = "templates"
+    TEMPLATE_CALL: str = "template"
+
+    def __init__(self, data: AttrDict):
+        """Initialise the solver."""
+        self._raw_templates: AttrDict = data.get_key(self.TEMPLATES_SECTION, AttrDict())
+        self._raw_data: AttrDict = data
+        self.resolved_templates: AttrDict
+        self.resolved_data: AttrDict
+        self._resolve()
+
+    def _resolve(self):
+        """Fill in template references and remove template definitions and calls."""
+        self.resolved_templates = AttrDict()
+        for key, value in self._raw_templates.items():
+            if not isinstance(value, dict):
+                raise exceptions.ModelError("Template definitions must be YAML blocks.")
+            self.resolved_templates[key] = self._resolve_template(key)
+        self.resolved_data = self._resolve_data(self._raw_data)
+
+    def _resolve_template(self, name: str, stack: None | set[str] = None) -> AttrDict:
+        """Resolves templates recursively.
+
+        Catches circular template definitions.
+        """
+        if stack is None:
+            stack = set()
+        elif name in stack:
+            raise exceptions.ModelError(
+                f"Circular template reference detected for '{name}'."
+            )
+        stack.add(name)
+
+        result = AttrDict()
+        raw_data = self._raw_templates[name]
+        if self.TEMPLATE_CALL in raw_data:
+            # Current template takes precedence when overriding values
+            inherited_name = raw_data[self.TEMPLATE_CALL]
+            if inherited_name in self.resolved_templates:
+                inherited_data = self.resolved_templates[inherited_name]
+            else:
+                inherited_data = self._resolve_template(inherited_name, stack)
+            result.union(inherited_data)
+
+        local_data = {k: raw_data[k] for k in raw_data.keys() - {self.TEMPLATE_CALL}}
+        result.union(local_data, allow_override=True)
+
+        stack.remove(name)
+        return result
+
+    def _resolve_data(self, section, level: int = 0):
+        if isinstance(section, dict):
+            if self.TEMPLATES_SECTION in section:
+                if level != 0:
+                    raise exceptions.ModelError(
```

Comment:
The idea is to have the schema fit what the model data should have _at the moment of inclusion_. This is why I removed the templates portion of the schema. For the future `pydantic` update, this would mean that templates are not part of the validation model, simplifying our case.

Reply:
More explanation: if the templates section reaches the moment where the schema is applied, it means that dictionary was not produced by our model_definition function. Thus, we should not trust it and fail.

Reply:
Given the developments in other PRs, the schema / `pydantic` would eventually catch this.
I added this because it feels like a useful error message, and it is within the 'scope' of the template functionality.